src/notification-banner/private/useNotificationBanners.jsx

```diff
@@ -0,0 +1,182 @@
+ *   - NotificationBanners - React Component - Renders a list of prioritized notification banners.
+ *
+ * @returns { NotificationBannerProvider, NotificationBanners }
+ */
+const useNotificationBanners = () => {
+  /** The registeredBanners ref maintains collection of banner ids and banner props registered to the  NotificationBannerProvider. */
+  const registeredBanners = React.useRef({});
+
+  /**
+   * The updateBannerState ref stores the update state function used to manage the banner rendered in the NotificationBanners component.
+   * This ties the state updates to the `useNotificationBanners` hook, while allowing the NotificationBanners to be rendered above or below
+   * the NotificationBannerProvider.
+   */
+  const updateBannerState = React.useRef();
+
+  const useNotificationBannerExports = React.useMemo(() => {
+    /**
+    * Adds the banner ID and props to the collection of registered banners. Once registered,
+    * it updates the NotificationBanner's state to render the new Banner.
+    *
+    * @param {UUID} bannerId - unique ID associated to the banner
+    * @param {Object} bannerProps - react props associated to the banner. See ../NotificationBanner's propTypes.
+    */
+    const registerNotificationBanner = (bannerId, bannerProps) => {
+      if (!bannerId) {
+        throw new Error('A banner cannot be registered without an identifier.');
+      }
+
+      const { variant } = bannerProps;
+
+      if (!registeredBanners.current[variant]) {
+        registeredBanners.current[variant] = {};
+      }
+
+      registeredBanners.current[variant][bannerId] = { key: bannerId, ...bannerProps };
+
+      if (updateBannerState.current) {
+        updateBannerState.current({ ...registeredBanners.current });
+      }
+    };
+
+    /**
+      * Removes the banner ID and props from the collection of registered banners. Once unregistered,
+      * it updates the NotificationBanner's state to remove the Banner from the list.
+      *
+      * @param {UUID} bannerId - unique ID associated to the banner
+      * @param {String} bannerVariant - the banner variant to remove banner from
+      */
+    const unregisterNotificationBanner = (bannerId, bannerVariant) => {
+      if (!bannerId || !bannerVariant) {
+        throw new Error('A banner cannot be unregistered without an identifier or banner variant.');
+      }
+
+      if (!registeredBanners.current[bannerVariant][bannerId]) {
+        return;
+      }
+
+      delete registeredBanners.current[bannerVariant][bannerId];
+
+      if (updateBannerState.current) {
+        updateBannerState.current({ ...registeredBanners.current });
+      }
+    };
+
+    return {
+      /**
+       * Provides the Banner Registration Context to its children.
+       */
+      NotificationBannerProvider: ({ children }) => ( // eslint-disable-line react/prop-types
+        <BannerRegistrationContext.Provider value={{ registerNotificationBanner, unregisterNotificationBanner }}>
```

Comment:
I believe this will still create a new value object with every update. You could move the value into a const outside the component (like to line 84) to avoid this.

Reply:
This hook is only ran on mount so I'm not sure it matters if this is a declared as a const or created in-line. I can update this though.

Reply:
The NotificationBannerProvider component is created once, but the NotificationBannerProvider is executed every time it gets rendered, and as written each execution will create a new value object. Moving the object to the outer scope ensures a consistent reference is used with each render.